transaction/core/src/input_rules.rs

```diff
@@ -53,6 +69,97 @@ impl InputRules {
                 return Err(InputRuleError::MissingRequiredOutput);
             }
         }
+        // Partial-Fill rules verification (MCIP #42)
+        if let Some(fractional_change) = self.fractional_change.as_ref() {
+            // There is a fractional change output. Let's try to unblind its amount.
+            let fractional_change_amount = fractional_change.reveal_amount()?;
+            // Let's check if there is a corresponding real change output.
+            let real_change = tx
+                .prefix
+                .outputs
+                .iter()
+                .find(|x| x.public_key == fractional_change.tx_out.public_key)
+                .ok_or(InputRuleError::MissingRealChangeOutput)?;
+            // Let's try to unblind its amount.
+            let real_change_amount =
+                try_reveal_amount(real_change, fractional_change.amount_shared_secret.as_ref())?;
+
+            // Check the bounds of the real change amount
+            if real_change_amount.token_id != fractional_change_amount.token_id {
+                return Err(InputRuleError::RealOutputTokenIdMismatch);
+            }
+            if real_change_amount.value > fractional_change_amount.value {
+                return Err(InputRuleError::RealOutputAmountExceedsFractional);
+            }
+            if self.max_allowed_change_value != 0
+                && real_change_amount.value > self.max_allowed_change_value
+            {
+                return Err(InputRuleError::RealChangeOutputAmountExceedsLimit);
+            }
+
+            // Compute fill-fraction num and denom. This is the fraction of the maximum
+            // possible trade that could have occurred which did occur. This
+            // calculation is used later when checking if real output amounts respected the
+            // fill fraction.
+            let fill_fraction_num =
+                (fractional_change_amount.value - real_change_amount.value) as u128;
+            let fill_fraction_denom = fractional_change_amount.value as u128;
+
+            // Verify fractional_outputs
+            for fractional_output in self.fractional_outputs.iter() {
```

Comment:
Yeah they just end up screwing themselves

This is a silly thing to do but I'm not sure if it should be an error.

I'm trying to keep in mind that we might add more stuff to `InputRules` later that interacts with existing fields, so it's possible that at a future revision it's okay for `self.fractional_outputs` to be empty because some other list is not empty.

Idk, maybe that's not good, maybe it should just be an error. I've tried to make a lot of forms of misuse be an error.

Reply:
I don't feel strongly - mostly wanted to test my understanding of that edgecase.